utilities/ideascale-importer/ideascale_importer/snapshot_importer/importer.py

```diff
@@ -103,10 +107,9 @@ async def _fetch_parameters(self):
             )
 
             row = await conn.fetchrow(
-                "SELECT slot_no FROM block WHERE time <= $1 ORDER BY time DESC LIMIT 1",
+                "SELECT MAX(slot_no) as slot_no FROM block WHERE time <= $1 AND slot_no IS NOT NULL",
```

Comment:
This will get the slot of the deadline,  but ONLY when the deadline has passed.
Which means the next line will fail by raising the exception.
Instead,  IF the deadline has not passed, and this returns None, it should just default to the latest slot number.
So we end up with 3 time bands.

1. Before the deadline.  (snapshot_tool_max_slot == latest_slot). Not Final.
2. After the deadline and latest_slot_time < snapshot_start (snapshot_tool_max_slot = slot just before deadline). Not final.
3. After the deadline and latest_slot_time >= snapshot_start (snapshot_tool_max_slot = slot just before deadline). Final.

Reply:
`max_time` is used to get the `max_slot` parameter the importer passes to snapshot_tool. This query is getting the slot_no at (at most) that time. This is independent of the deadline.

The snapshot deadline is fetched from the event in the DB and stored in the `Importer.snapshot_start_time` [here](https://github.com/input-output-hk/catalyst-core/pull/307/files/41c99240f7a13ea80a30ded7355618b6672bc265#diff-8c922024c086e6e0c07f0c8d115ec575671ba9bb385b86b6e69ab41bbd08aa2cR146). This is then used [here](https://github.com/input-output-hk/catalyst-core/pull/307/files/41c99240f7a13ea80a30ded7355618b6672bc265#diff-8c922024c086e6e0c07f0c8d115ec575671ba9bb385b86b6e69ab41bbd08aa2cR269) to decide whether the snapshot is final.

Reply:
As discussed in slack,  this is an error.  There are 2 times in the DB.
`registration_snapshot_time` this is the deadline, and the slot just PRIOR to this time is the `max_slot` passed to the snapshot tool.  But we need to make sure this time has past (as far as dbsync is concerned), because if it hasn't we want to use the latest slot we have.

`snapshot_start` is named this because its when we used to run snapshot when there was only 1.  But now it means snapshot is final if we ran one and the latest slot time was after this time.